File: insert_middle.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
def insert_middle(head, data):
    if not head:
        return Node(data)
    slow = head
    fast = head
    prev = None
    while fast and fast.next:
        prev = slow
        slow = slow.next
        fast = fast.next.next
    new_node = Node(data)
    new_node.next = slow
    prev.next = new_node
    return head
def display(head):
    temp = head
    while temp:
        print(temp.data, "->", end=" ")
        temp = temp.next
    print("NULL")
if __name__ == "__main__":
    head = None
    n = int(input("Enter number of elements: "))
    for _ in range(n):
        val = int(input("Enter value: "))
        node = Node(val)
        node.next = head
        head = node
    head = insert_middle(head, 99)
    display(head)

File: insert_middle.c
#include <stdio.h>
#include <stdlib.h>
struct Node {
    int data;
    struct Node* next;
};
struct Node* insertMiddle(struct Node* head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    if (head == NULL) {
        newNode->next = NULL;
        return newNode;
    }
    struct Node* slow = head;
    struct Node* fast = head;
    struct Node* prev = NULL;
    while (fast != NULL && fast->next != NULL) {
        prev = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    newNode->next = slow;
    prev->next = newNode;
    return head;
}
void display(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}
int main() {
    struct Node* head = NULL;
    struct Node* temp;
    int n, val;
    printf("Enter number of elements: ");
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
        printf("Enter value: ");
        scanf("%d", &val);
        newNode->data = val;
        newNode->next = head;
        head = newNode;
    }
    head = insertMiddle(head, 99);
    display(head);
    return 0;
}
